cpp/src/bitmask/bit_container.cuh

```diff
@@ -0,0 +1,239 @@
+   */
+  template <typename T>
+  __host__ __device__
+  bool IsValid(T record_idx) const {
+    return is_valid(valid_, record_idx);
+  }
+
+  /**
+   * Set a bit
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __host__ __device__
+  void SetBit(T record_idx) {
+    int rec = which_word(record_idx);
+    int bit = which_bit(record_idx);
+
+    valid_[rec] = valid_[rec] | (1U << bit);
+  }
+
+  /**
+   * Clear a bit
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __host__ __device__
+  void ClearBit(T record_idx) {
+    int rec = which_word(record_idx);
+    int bit = which_bit(record_idx);
+
+    valid_[rec] = valid_[rec] & (~(1U << bit));
+  }
+
+  /**
+   * Getter for valid
+   *
+   * @return pointer to valid bitmask
+   */
+  __host__ __device__
+  bit_container_t *GetValid() const {
+    return valid_;
+  }
+
+  /**
+   * Get length
+   *
+   * @return length of bitmask in bits
+   */
+  __host__ __device__
+  int Length() const {
+    return bitlength_;
+  }
+
+  /**
+   * Get the number of words in this bit mask
+   *
+   * @return the number of words
+   */
+  inline
+  __host__ __device__
+  int NumWords() const {
+    return num_words(bitlength_);
+  }
+
+private:
+  /**
+   *   array of entries containing the bitmask
+   */
+};
+
+
+/* ---------------------------------------------------------------------------- */
+/**
+ * @Synopsis  Class for managing bit masks on the device
+ */
+/* ---------------------------------------------------------------------------- */
+class BitContainer {
+public:
+   __device__ BitContainer(bit_container_t *valid, int bitlength): valid_(valid), bitlength_(bitlength) {}
+   __device__ BitContainer(const BitContainer &other): valid_(other.valid_), bitlength_(other.bitlength_) {}
+
+  /**
+   * Check to see if a record is Valid (aka not null)
+   *
+   * @param[in] record_idx   the record index to check
+   *
+   * @return  true if record is valid, false if record is null
+   */
+  template <typename T>
+  __device__ bool IsValid(T record_idx) const {
+    return bit_container::is_valid(record_idx);
+  }
+
+  /**
+   * Set a bit (not thread-save)
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __device__ void SetBitUnsafe(T record_idx) {
+    bit_container::set_bit_unsafe(valid_, record_idx);
+  }
+
+
+  /**
+   * Clear a bit (not thread-safe)
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __device__ void ClearBitUnsafe(T record_idx) {
+    bit_container::clear_bit_unsafe(record_idx);
+  }
+
+  /**
+   * Set a bit in a thread-safe manner
+   *
+   * @param[in] record_idx    the record index
+   */
+  template <typename T>
+  __device__ void SetBit(T record_idx) {
+    int rec = bit_container::which_word(record_idx);
+    int bit = bit_container::which_bit(record_idx);
+
+    atomicOr( &valid_[rec], (1U << bit));
```

Comment:
This is more like it. What is the version in Util? Why does Util duplicate all of this functionality?